### ISSUE TITLE
[JENKINS-44820] Parallel branches with no stage should not be displayed in Blue Ocean

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#4201]
<p>With the following pipeline:</p>

<pre>
stage(<span class="code-quote">"begin"</span>) { println <span class="code-quote">"begin"</span> }

parallel (
    <span class="code-quote">"no_stage"</span>: { println <span class="code-quote">"no stage"</span> },
    
    <span class="code-quote">"": { println "</span>no name" },
    
    <span class="code-quote">"with_stage"</span>: { stage(<span class="code-quote">"my_stage"</span>) { println <span class="code-quote">"named stage"</span> } }
)

stage(<span class="code-quote">"end"</span>) { println <span class="code-quote">"end"</span> }
</pre>
<p>we are getting the following display in BlueOcean:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/38443/38443_blueocean.png" style="border: 0px solid black" /></span></p>

<p>But in the parallel step, only the third branch has a <b>stage</b> defined, and therefore only this one is expected to be displayed. Unfortunately all the branches are displayed even if not containing any stage.<br/>
Additionally, the name displayed for the third branch is the <b>branch name</b> instead of the <b>stage name</b> as one would have thought.</p>

<p>On the other hand, in the Pipeline view it is displayed as expected:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/38442/38442_stages.png" style="border: 0px solid black" /></span></p>

<p>Use case: technical steps which need to run in parallel to maintain shared resources, advance timeout monitoring, etc ... doesn't have to pollute the Blue Ocean view. </p>

<p>May be possible to add a boolean option to the parallel step not to use branch name as default display name in Blue Ocean, and eventually make branch with no stage invisible ?</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jguigui">jguigui</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44820" target="_blank">Parallel branches with no stage should not be displayed in Blue Ocean</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>component(s)</b>: blueocean-plugin
<li><b>label(s)</b>: blueocean, parallel, stage, ui
<li><b>resolution</b>: Duplicate
<li><b>resolved</b>: 2017-11-29T04:44:52+00:00
<li><b>votes</b>: 3
<li><b>watchers</b>: 4
<li><b>imported</b>: 2025-12-01
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>With the following pipeline:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage(<span class="code-quote">"begin"</span>) { println <span class="code-quote">"begin"</span> }

parallel (
    <span class="code-quote">"no_stage"</span>: { println <span class="code-quote">"no stage"</span> },
    
    <span class="code-quote">"": { println "</span>no name" },
    
    <span class="code-quote">"with_stage"</span>: { stage(<span class="code-quote">"my_stage"</span>) { println <span class="code-quote">"named stage"</span> } }
)

stage(<span class="code-quote">"end"</span>) { println <span class="code-quote">"end"</span> }
</pre>
</div></div>
<p>we are getting the following display in BlueOcean:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/38443/38443_blueocean.png" style="border: 0px solid black" /></span></p>

<p>But in the parallel step, only the third branch has a <b>stage</b> defined, and therefore only this one is expected to be displayed. Unfortunately all the branches are displayed even if not containing any stage.
Additionally, the name displayed for the third branch is the <b>branch name</b> instead of the <b>stage name</b> as one would have thought.</p>

<p>On the other hand, in the Pipeline view it is displayed as expected:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/38442/38442_stages.png" style="border: 0px solid black" /></span></p>

<p>Use case: technical steps which need to run in parallel to maintain shared resources, advance timeout monitoring, etc ... doesn't have to pollute the Blue Ocean view. </p>

<p>May be possible to add a boolean option to the parallel step not to use branch name as default display name in Blue Ocean, and eventually make branch with no stage invisible ?</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
jenkins 2.64<br/>
blue ocean 1.0.1
```
</details>
<details><summary><i>6 attachments</i></summary>

- [[JENKINS-44820] Parallel branches with no stage should not be displayed in Blue Ocean - Jenkins JIRA 2017-06-22 14-18-59.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38582/%5BJENKINS-44820%5D%20Parallel%20branches%20with%20no%20stage%20should%20not%20be%20displayed%20in%20Blue%20Ocean%20-%20Jenkins%20JIRA%202017-06-22%2014-18-59.png)
  > ![[JENKINS-44820] Parallel branches with no stage should not be displayed in Blue Ocean - Jenkins JIRA 2017-06-22 14-18-59.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38582/%5BJENKINS-44820%5D%20Parallel%20branches%20with%20no%20stage%20should%20not%20be%20displayed%20in%20Blue%20Ocean%20-%20Jenkins%20JIRA%202017-06-22%2014-18-59.png)
- [blueocean.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/38443/blueocean.png)
  > ![blueocean.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/38443/blueocean.png)
- [image-2017-06-22-12-33-47-644.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38583/image-2017-06-22-12-33-47-644.png)
  > ![image-2017-06-22-12-33-47-644.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38583/image-2017-06-22-12-33-47-644.png)
- [image-2017-06-22-12-47-13-324.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/38584/image-2017-06-22-12-47-13-324.png)
  > ![image-2017-06-22-12-47-13-324.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/38584/image-2017-06-22-12-47-13-324.png)
- [image-2017-06-22-12-53-45-573.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38585/image-2017-06-22-12-53-45-573.png)
  > ![image-2017-06-22-12-53-45-573.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38585/image-2017-06-22-12-53-45-573.png)
- [stages.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38442/stages.png)
  > ![stages.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38442/stages.png)
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote link associated with this issue:_


  - [Page](https://wiki.jenkins.io/pages/viewpage.action?pageId=135469610)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44820?focusedId=304327&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304327">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jguigui" class="user-hover" rel="jguigui">jguigui</a> this is related to a bunch of problems around sophisticated Pipelines.</p>

<p>I am in the process of drafting a RFC to rationalise parallel usage.</p>

<p>For example, in Declarative 1.2 we've introduced the following syntax called "Parallel Stages"</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage(<span class="code-quote">'foo'</span>) {
    parallel {
        stage(<span class="code-quote">'Firefox'</span>) {
            steps {
                echo <span class="code-quote">"First branch"</span>
            }
        }
        stage(<span class="code-quote">'Chrome'</span>) {
            steps {
                echo <span class="code-quote">"Second branch"</span>
            }
        }
    }
}
</pre>
</div></div>

<p>The advantage here is that we now have different syntaxes for parallel execution and parallel stages. </p>

<p>For reference the current syntax looks like:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage(<span class="code-quote">'foo'</span>) {
    parallel(
		<span class="code-quote">'copy-1'</span>: {
			... copy some files ...
		},
		<span class="code-quote">'copy-2'</span>: {
			... copy some files ...
		},
		<span class="code-quote">'some other process'</span>: {
			 parallel(
				<span class="code-quote">'branch-1'</span>: {
					... run process 1 ...
				},
				<span class="code-quote">'branch-2'</span>: {
					... run process 2 ...
				}
			)
		}
	)
}
</pre>
</div></div>

<p>With the syntax above its possible to next parallels infinitely and this is something Blue Ocean isn't likely to support.</p>

<p>What I would like to do is:</p>
<ul>
	<li>Make the new Parallel Stage syntax the only way to visualise parallel</li>
	<li>Parallel executions get folded into the nearest "stage" and their steps appear in the step listing according to their execution order.</li>
</ul>


<p>This is still some time away but it would cover the case you have outlined here.</p>

<p>Let me know what you think</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jguigui" class="user-hover" rel="jguigui">jguigui</a> this is related to a bunch of problems around sophisticated Pipelines.</p>

<p>I am in the process of drafting a RFC to rationalise parallel usage.</p>

<p>For example, in Declarative 1.2 we've introduced the following syntax called "Parallel Stages"</p>


<pre>
stage(<span class="code-quote">'foo'</span>) {
    parallel {
stage(<span class="code-quote">'Firefox'</span>) {
    steps {
echo <span class="code-quote">"First branch"</span>
    }
}
stage(<span class="code-quote">'Chrome'</span>) {
    steps {
echo <span class="code-quote">"Second branch"</span>
    }
}
    }
}
</pre>

<p>The advantage here is that we now have different syntaxes for parallel execution and parallel stages. </p>

<p>For reference the current syntax looks like:</p>

<pre>
stage(<span class="code-quote">'foo'</span>) {
    parallel(
		<span class="code-quote">'copy-1'</span>: {
			... copy some files ...
		},
		<span class="code-quote">'copy-2'</span>: {
			... copy some files ...
		},
		<span class="code-quote">'some other process'</span>: {
			 parallel(
				<span class="code-quote">'branch-1'</span>: {
					... run process 1 ...
				},
				<span class="code-quote">'branch-2'</span>: {
					... run process 2 ...
				}
			)
		}
	)
}
</pre>

<p>With the syntax above its possible to next parallels infinitely and this is something Blue Ocean isn't likely to support.</p>

<p>What I would like to do is:</p>
<ul>
	<li>Make the new Parallel Stage syntax the only way to visualise parallel</li>
	<li>Parallel executions get folded into the nearest "stage" and their steps appear in the step listing according to their execution order.</li>
</ul>


<p>This is still some time away but it would cover the case you have outlined here.</p>

<p>Let me know what you think</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jguigui:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44820?focusedId=304349&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304349">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi James,</p>

<p>I think there are two orthogonal concepts here:</p>
<ul>
	<li>the parallel <b>branches</b> used to run in parallel (on different cpu for instance) several tasks</li>
	<li>the <b>stage</b> used for display purpose</li>
</ul>


<p>To be flexible and cover all the use cases, we need to have these two concept independent.
Typically we may want:</p>
<ol>
	<li>in a same branch have sequential stages</li>
	<li>in a same branch have nested stages</li>
	<li>having branch without stage (aka no display)</li>
</ol>


<p>Looks like your RFC is covering cases 1. and 2. but not 3.</p>

<p>Today limitation with current implementation are:</p>
<ul>
	<li>the parallel branch is creating a stage with this branch name as the stage name</li>
	<li>branches with no stages are displayed (linked to first point here above)</li>
	<li>nested stages are not displayed</li>
</ul>


<p>Having the branch name (used in the log) different from the stage name (used in Blue Ocean display) is useful too.</p>

<p>Regarding the representation of nested stages I like your suggestion. There is similar discussion on <a href="https://issues.jenkins.io/browse/JENKINS-38442" title="View sequential stages in the pipeline visualization graph" class="issue-link" data-issue-key="JENKINS-38442"><del>JENKINS-38442</del></a>.</p>

<p>Last point I can think about, is how to represent stage with <b>same name</b> in a parallel step. Typically when lot of branches are generated by a for loop. I'm not sure of the best option:</p>
<ul>
	<li>either aggregate them in one display (with eventually a badge indicating the number of stages with same name)</li>
	<li>either display each of them</li>
</ul>



<p>Jean-Paul</p>
</pre>
</details>

</ul>
</details>

<p>Hi James,</p>

<p>I think there are two orthogonal concepts here:</p>
<ul>
	<li>the parallel <b>branches</b> used to run in parallel (on different cpu for instance) several tasks</li>
	<li>the <b>stage</b> used for display purpose</li>
</ul>


<p>To be flexible and cover all the use cases, we need to have these two concept independent.<br/>
Typically we may want:</p>
<ol>
	<li>in a same branch have sequential stages</li>
	<li>in a same branch have nested stages</li>
	<li>having branch without stage (aka no display)</li>
</ol>


<p>Looks like your RFC is covering cases 1. and 2. but not 3.</p>

<p>Today limitation with current implementation are:</p>
<ul>
	<li>the parallel branch is creating a stage with this branch name as the stage name</li>
	<li>branches with no stages are displayed (linked to first point here above)</li>
	<li>nested stages are not displayed</li>
</ul>


<p>Having the branch name (used in the log) different from the stage name (used in Blue Ocean display) is useful too.</p>

<p>Regarding the representation of nested stages I like your suggestion. There is similar discussion on <a href="https://issues.jenkins.io/browse/JENKINS-38442" title="View sequential stages in the pipeline visualization graph" class="issue-link" data-issue-key="JENKINS-38442"><del>JENKINS-38442</del></a>.</p>

<p>Last point I can think about, is how to represent stage with <b>same name</b> in a parallel step. Typically when lot of branches are generated by a for loop. I'm not sure of the best option:</p>
<ul>
	<li>either aggregate them in one display (with eventually a badge indicating the number of stages with same name)</li>
	<li>either display each of them</li>
</ul>



<p>Jean-Paul</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44820?focusedId=304350&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304350">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jguigui" class="user-hover" rel="jguigui">jguigui</a> </p>

<blockquote>
<p>having branch without stage (aka no display)</p></blockquote>

<p>These will still need to be in a stage but you would not see the "parallels" in the visualisation. You need the enclosing stage for us to display the steps within the parallel.</p>

<p>In your example, if we made these changes, your visualisation would look like:
 <span class="image-wrap" style=""><a id="38582_thumb" href="https://issues.jenkins.io/secure/attachment/38582/38582_%5BJENKINS-44820%5D+Parallel+branches+with+no+stage+should+not+be+displayed+in+Blue+Ocean+-+Jenkins+JIRA+2017-06-22+14-18-59.png" title="[JENKINS-44820] Parallel branches with no stage should not be displayed in Blue Ocean - Jenkins JIRA 2017-06-22 14-18-59.png" file-preview-type="image" file-preview-id="38582" file-preview-title="[JENKINS-44820] Parallel branches with no stage should not be displayed in Blue Ocean - Jenkins JIRA 2017-06-22 14-18-59.png"><img src="https://issues.jenkins.io/secure/thumbnail/38582/_thumb_38582.png" style="border: 0px solid black" role="presentation"/></a></span> </p>

<p>It is also highly unlikely we will be ever able to visualise infinitely nested parallels. This proposal would not attempt to address that use case.</p>

<p>Thanks for the response <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/thumbs_up.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jguigui" class="user-hover" rel="jguigui">jguigui</a> </p>

<blockquote>
<p>having branch without stage (aka no display)</p></blockquote>

<p>These will still need to be in a stage but you would not see the "parallels" in the visualisation. You need the enclosing stage for us to display the steps within the parallel.</p>

<p>In your example, if we made these changes, your visualisation would look like:<br/>
 <span class="image-wrap" style=""><a id="38582_thumb" href="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38582/%5BJENKINS-44820%5D%20Parallel%20branches%20with%20no%20stage%20should%20not%20be%20displayed%20in%20Blue%20Ocean%20-%20Jenkins%20JIRA%202017-06-22%2014-18-59.png" title="[JENKINS-44820] Parallel branches with no stage should not be displayed in Blue Ocean - Jenkins JIRA 2017-06-22 14-18-59.png" file-preview-type="image" file-preview-id="38582" file-preview-title="[JENKINS-44820] Parallel branches with no stage should not be displayed in Blue Ocean - Jenkins JIRA 2017-06-22 14-18-59.png"><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38582/%5BJENKINS-44820%5D%20Parallel%20branches%20with%20no%20stage%20should%20not%20be%20displayed%20in%20Blue%20Ocean%20-%20Jenkins%20JIRA%202017-06-22%2014-18-59.png" style="border: 0px solid black" role="presentation"/></a></span> </p>

<p>It is also highly unlikely we will be ever able to visualise infinitely nested parallels. This proposal would not attempt to address that use case.</p>

<p>Thanks for the response <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/thumbs_up.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jguigui:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44820?focusedId=304352&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304352">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi James,</p>

<p>If I understood properly, you mean that the following pipeline:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage(<span class="code-quote">"begin"</span>) { println <span class="code-quote">"begin"</span> }

stage(<span class="code-quote">'stage_parallel'</span>) {
    parallel {
        stage(<span class="code-quote">'stage_p1'</span>) {
            steps {
                echo <span class="code-quote">"First branch"</span>
            }
        }
        stage(<span class="code-quote">''stage_p2'</span>) {
            steps {
                echo <span class="code-quote">"Second branch"</span>
            }
        }
    }
}

stage(<span class="code-quote">"end"</span>) { println <span class="code-quote">"end"</span> }

</pre>
</div></div>
<p>will not display the nested stage_p1 and stage_p2 and generate this:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/38583/38583_image-2017-06-22-12-33-47-644.png" style="border: 0px solid black" /></span></p>

<p> </p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>Hi James,</p>

<p>If I understood properly, you mean that the following pipeline:</p>

<pre>
stage(<span class="code-quote">"begin"</span>) { println <span class="code-quote">"begin"</span> }

stage(<span class="code-quote">'stage_parallel'</span>) {
    parallel {
stage(<span class="code-quote">'stage_p1'</span>) {
    steps {
echo <span class="code-quote">"First branch"</span>
    }
}
stage(<span class="code-quote">''stage_p2'</span>) {
    steps {
echo <span class="code-quote">"Second branch"</span>
    }
}
    }
}

stage(<span class="code-quote">"end"</span>) { println <span class="code-quote">"end"</span> }

</pre>
<p>will not display the nested stage_p1 and stage_p2 and generate this:</p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38583/image-2017-06-22-12-33-47-644.png" style="border: 0px solid black" /></span></p>

<p> </p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44820?focusedId=304353&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304353">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jguigui" class="user-hover" rel="jguigui">jguigui</a> that is correct</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jguigui" class="user-hover" rel="jguigui">jguigui</a> that is correct</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44820?focusedId=304354&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304354">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jguigui" class="user-hover" rel="jguigui">jguigui</a> actually, I spoke too soon. This pipeline:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage(<span class="code-quote">"begin"</span>) { println <span class="code-quote">"begin"</span> }

stage(<span class="code-quote">'stage_parallel'</span>) {
    parallel {
        <span class="code-quote">'stage_p1'</span>: {
                echo <span class="code-quote">"First branch"</span>
        }
         <span class="code-quote">'stage_p2'</span>: {
                echo <span class="code-quote">"Second branch"</span>
        }
    }
}

stage(<span class="code-quote">"end"</span>) { println <span class="code-quote">"end"</span> }
</pre>
</div></div>

<p>Would produce:
 <span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/38583/38583_image-2017-06-22-12-33-47-644.png" style="border: 0px solid black" /></span> </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jguigui" class="user-hover" rel="jguigui">jguigui</a> actually, I spoke too soon. This pipeline:</p>


<pre>
stage(<span class="code-quote">"begin"</span>) { println <span class="code-quote">"begin"</span> }

stage(<span class="code-quote">'stage_parallel'</span>) {
    parallel {
<span class="code-quote">'stage_p1'</span>: {
echo <span class="code-quote">"First branch"</span>
}
 <span class="code-quote">'stage_p2'</span>: {
echo <span class="code-quote">"Second branch"</span>
}
    }
}

stage(<span class="code-quote">"end"</span>) { println <span class="code-quote">"end"</span> }
</pre>

<p>Would produce:<br/>
 <span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38583/image-2017-06-22-12-33-47-644.png" style="border: 0px solid black" /></span> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jguigui:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44820?focusedId=304356&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304356">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>And the following pipeline with some branches not inside any stage:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage(<span class="code-quote">"begin"</span>) { println <span class="code-quote">"begin"</span> }

stage(<span class="code-quote">'stage_parallel'</span>) {
    parallel {
        stage(<span class="code-quote">'stage_p1'</span>) {
            steps {
                echo <span class="code-quote">"First branch"</span>
            }
        }
        stage(<span class="code-quote">'stage_p2'</span>) {
            steps {
                echo <span class="code-quote">"Second branch"</span>
            }
        }
        <span class="code-quote">'invisible_branch_p3'</span>: {
                echo <span class="code-quote">"Third branch"</span>
        }
        <span class="code-quote">'invisible_branch_p4'</span>: {
                echo <span class="code-quote">"Fourth branch"</span>
        }
    }
}

stage(<span class="code-quote">"end"</span>) { println <span class="code-quote">"end"</span> }
</pre>
</div></div>
<p>would produce with the proposed RFC:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/38585/38585_image-2017-06-22-12-53-45-573.png" style="border: 0px solid black" /></span></p>

<p> </p>

<p>If it's the case, I've my use cases covered.</p>

<p>Agree with you, multiple nested case will be difficult to display.</p>

<p>Thanks a lot,</p>

<p>Jean-Paul</p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>And the following pipeline with some branches not inside any stage:</p>

<pre>
stage(<span class="code-quote">"begin"</span>) { println <span class="code-quote">"begin"</span> }

stage(<span class="code-quote">'stage_parallel'</span>) {
    parallel {
stage(<span class="code-quote">'stage_p1'</span>) {
    steps {
echo <span class="code-quote">"First branch"</span>
    }
}
stage(<span class="code-quote">'stage_p2'</span>) {
    steps {
echo <span class="code-quote">"Second branch"</span>
    }
}
<span class="code-quote">'invisible_branch_p3'</span>: {
echo <span class="code-quote">"Third branch"</span>
}
<span class="code-quote">'invisible_branch_p4'</span>: {
echo <span class="code-quote">"Fourth branch"</span>
}
    }
}

stage(<span class="code-quote">"end"</span>) { println <span class="code-quote">"end"</span> }
</pre>
<p>would produce with the proposed RFC:</p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38585/image-2017-06-22-12-53-45-573.png" style="border: 0px solid black" /></span></p>

<p> </p>

<p>If it's the case, I've my use cases covered.</p>

<p>Agree with you, multiple nested case will be difficult to display.</p>

<p>Thanks a lot,</p>

<p>Jean-Paul</p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44820?focusedId=304701&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304701">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jguigui" class="user-hover" rel="jguigui">jguigui</a> you wouldn't be able to put both parallel syntaxes within the same stage with the proposal as it stands <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/sad.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jguigui" class="user-hover" rel="jguigui">jguigui</a> you wouldn't be able to put both parallel syntaxes within the same stage with the proposal as it stands <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/sad.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jguigui:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44820?focusedId=304727&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304727">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I hope you may reconsider the RFC. Definitely by splitting the execution (branch) and the display (stage) concepts, users will be given a much better flexibility. </p>
</pre>
</details>

</ul>
</details>

<p>I hope you may reconsider the RFC. Definitely by splitting the execution (branch) and the display (stage) concepts, users will be given a much better flexibility. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44820?focusedId=304858&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304858">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jguigui" class="user-hover" rel="jguigui">jguigui</a> its still in the planning stages so I will consider when driving a decision on this one <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/thumbs_up.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jguigui" class="user-hover" rel="jguigui">jguigui</a> its still in the planning stages so I will consider when driving a decision on this one <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/thumbs_up.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jguigui:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44820?focusedId=304862&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304862">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Great ! Thank you very much <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/thumbs_up.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>Jean-Paul</p>
</pre>
</details>

</ul>
</details>

<p>Great ! Thank you very much <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/thumbs_up.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>Jean-Paul</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44820?focusedId=321170&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-321170">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Merging this within <a href="https://issues.jenkins.io/browse/JENKINS-47799" title="Blue Ocean shouldn't visualize the parallel keyword of a scripted pipeline directly" class="issue-link" data-issue-key="JENKINS-47799">JENKINS-47799</a> as it will likely be the solution we are heading towards.</p>
</pre>
</details>

</ul>
</details>

<p>Merging this within <a href="https://issues.jenkins.io/browse/JENKINS-47799" title="Blue Ocean shouldn't visualize the parallel keyword of a scripted pipeline directly" class="issue-link" data-issue-key="JENKINS-47799">JENKINS-47799</a> as it will likely be the solution we are heading towards.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `duplicates` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-47799%22">JENKINS-47799</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-38442%22">JENKINS-38442</a>]</i>
